### PR TITLE
User Interface Translation

### PR DESCRIPTION
**Pull request by rafaelbsd**
Starting to work on translation

### CODE REVIEW

**Comment by endelwar** [comment]
I'm looking forward for this!

**Comment by rafaelbsd** [comment]
Hi there... 

I commit some files with partial tranlation ..... 

I´m still working on that but the most used stuff are already in the translation. 

Tks n Rgds! 

--- 
Rafael Mentz Aquino 
LK6 Soluções em TI 
Rua Domingos de Almeida, 135 sala 1102 
Centro - Novo Hamburgo - RS 
(51) 3065-6001 - 3035-6997 - 9999-7030 
www.lk6.com.br 

----- Mensagem original -----

**Comment by endelwar** [comment]
That's already a good work Rafael.

I think that it would be better if each translation stays in an array, like this:

```php
/* language/en.php */
$lang = array(
	'username' => 'User',
	'password' => 'Password',
	'cannot_read_conf' => "Cannot read conf.php - please create it by copying conf.php.example and modifying the parameters to suit."
);

return $lang;
```
and in function.php language are loaded like this:

```php
$lang = require_once(__DIR__ . DIRECTORY_SEPARATOR . 'languages' . DIRECTORY_SEPARATOR . LANG.'.php');
```
so that a translation can be reached using `$lang['username']` , or better creating a function to access translation:
```php
function _($string)
{
	$lang = require_once(__DIR__ . DIRECTORY_SEPARATOR . 'languages' . DIRECTORY_SEPARATOR . LANG.'.php');
	if (isset($lang[$string])) {
		return $lang[$string];
	}
	return null;
}
```

**Comment by stefaweb** [comment]
Hello!
Do you need help for French translation?
Are the i18 master files available?

**Comment by rafaelbsd** [comment]
Hi!

Yes. As a matter of fact, my idea is to create a file for each langague, so fell free to create mailwatch/mailscanner/langagues/fr.php and translate whats already converted. The only thing that I recomend, to run away of special caracthers problems (like portuguese and french have) is to use html code: á = &aacute; .

**Comment by rafaelbsd** [comment]
Hi endelwar, I just saw your comments.

I´ll work on that and replace my commits.

**Comment by stefaweb** [comment]
"special characters"

Il would love that you implement a filter for that. It is so boring to have to &aacute; all special characters in the language file. ;-) But will do if needed.

**Comment by rafaelbsd** [comment]
Oh.. .When I said á = á, I meant á = "\&aacute ;". But after I write, this page translated...  ;-)

**Comment by stefaweb** [comment]
Understood. ;-)

What is and where:
define('OTHERS03','Otherd');
define('SECONDS05','segunds');

**Comment by rafaelbsd** [comment]
Let me first apply the changes sugested by endelwar..... Because that will change the tranlation file format (will use array, instead of variables...). As soon as I commit the changes I´ll let you know.

**Comment by stefaweb** [comment]
Damed. Already done.
.
.
.
<?php

// 01-login.php
define('USERNAME01','Utilisateur');
define('PASSWORD01','Mot de passe');

// 03-functions.php
$MESSAGE01_03="Impossible de lire le fichier conf.php - Cr&eacute;ez un fichier de configuration conf.php.example et modifiez les param&egrave;tres de configuration comme n&eacute;cessaire.";
define('JUMPMESSAGE03','Aller au message :');
define('CUSER03','Utilisateur');
define('CST03','Heure syst&egrave;me');
define('COLORCODES03','Codes couleur');
define('CCODES01_03','Contenu dangereux/Infect&eacute;');
define('CCODES02_03','Liste Blanche');
define('CCODES03_03','Liste Noire');
define('CCODES04_03','Non v&eacute;rifi&eacute;');
define('CCODES05_03','Bon');
define('FREEDSPACE03','Espace disque libre');
define('TODAYSTOTALS03','Total aujourd\'hui');
define('PROCESSED03','Trait&eacute;');
define('CLEAN03','Bon');
define('VIRUSES03','Viruss&eacute;');
define('BLOCKEDFILES03','Fichier bloqu&eacute;');
define('OTHERS03','Otherd');
define('HSCOSPAM03','Spam &agrave; score tr&egrave;s haut');
define('HSCOMCP03','MCP &agrave; score tr&egrave;s haut');
define('RECENTMESSAGES03','Messages r&eacute;cent');
define('LISTS03','Listes');
define('QUARANTINE03','Quarantaine');
define('DATETIME03','Date/Heure');
define('FROM03','De');
define('TO03','A');
define('SIZE03','Taille');
define('SUBJECT03','Sujet');
define('SASCORE03','Score SA');
define('MCPSCORE03','Score MCP');

// 04-detail.php
define('RECEIVEDON04','Re&ccedil;u le');
define('RECEIVEDBY04','Re&ccedil;u par');
define('RECEIVEDFROM04','Re&ccedil;u de');
define('RECEIVEDVIA04','Re&ccedil;u via');
define('MSGHEADERS04','Headers du message');
define('FROM04','De');
define('TO04','A');
define('SIZE04','Taille');
define('SUBJECT04','Sujet');
define('HDRANTIVIRUS04','Anti-Virus/Protection contre les fichiers dangeureux');
define('BLKFILE04','Fichier bloqu&eacute;');
define('OTHERINFEC04','Autre infection');
define('OTHERINFEC04','Autre infection');
define('HSCOSPAM04','Spam &agrave; score tr&egrave;s haut');
define('LISTEDRBL04','List&eacute; en RBL');
define('SPAMWL04','SPAM en Liste Blanche');
define('SPAMBL04','SPAM en Liste Noire');
define('SAAUTOLEARN04','Auto aprentissage Spamassassin');
define('SASCORE04','Spamassassin Score');
define('SPAMREP04','Rapport de Spam');
define('HDRMCP04','Message Content Protection (MCP)');
define('HIGHSCOMCP04','High Scoring MCP');
define('MCPWL04','MCP en Liste Blanche');
define('MCPBL04','MCP en Liste Noire');
define('MCPSCORE04','Score MCP');
define('MCPREP04','Rapport MCP');
define('IPADDRESS04','Adresse IP');
define('COUNTRY04','Pays');
define('ALL04','Tous');
define('ADDWL04','Ajouter &agrave; la Liste Blanche');
define('ADDBL04','Ajouter &agrave; la Liste Noire');
define('RELEASE04','Lib&eacute;rer');
define('DELETE04','Supprimer');
define('SALEARN04','Apprentissage SA');
define('FILE04','Fichier');
define('TYPE04','Type');
define('PATH04','Chemin');
define('DANG04','Dangereux');
define('ALTRECIP04','Destinataire alternatif');
define('SUBMIT04','Soumettre');
define('ACTIONS04','Action(s)');
define('QUARCMDRES04','Quarantine Command Results');
define('RESULTMSGS04','Result Messages');

// 05-status.php
define('LAST05','Derniers');
define('MESSAGES05','Messages');
define('REFEVERY05','Rafraichir tous les');
define('SECONDS05','segunds');


// 06 - viewmail.php

define('MSGVIEWER06','Message Viewer');
define('RELEASEMSG06','Lib&eacute;rer ce message');
define('DELETEMSG06','Supprimer ce message');
define('ACTIONS06','Actions');
define('DATE06','Date :');
define('FROM06','De :');
define('TO06','A :');
define('SUBJECT06','Sujet :');

?>

**Comment by rafaelbsd** [comment]
Ahahaha... 
Whell, you can already use it. I´m using here too... My users are very happy... ehehehe

**Comment by rafaelbsd** [comment]
BTW, nice work! eheheh

**Comment by endelwar** [comment]
Let's keep accented letters and special chars in translation files (save them as utf-8), any format conversion will be done in `_` function

**Comment by stefaweb** [comment]
Very funny. When I past the code all the "&aacute ;" are automatically translated by git...

**Comment by endelwar** [comment]
@stefaweb you are too fast ;)

**Comment by endelwar** [comment]
@rafaelbsd also be aware that some translations are the same (USERNAME01 == CUSER03) and drop the number at the end of string name

**Comment by stefaweb** [comment]
Be careful with translation.
Depending of the context, French language use different word where English use only one uniq word.
Exemple : clean = "propre" or "bon".

**Comment by rafaelbsd** [comment]
Yes... in portuguese too.

**Comment by endelwar** [comment]
Ok then, let's keep two separate strings (also in italian is the same in some occasion)

**Comment by rafaelbsd** [comment]
I started using the variable names with a number, to reference the file.... I prefer to repeat a translation instead of write a wrong expression. I just finished change the files to the new format, and wlll commit now.

**Comment by rafaelbsd** [comment]
Hey Stephany, I already ajust you fr.php file to the new format, to save you work. Already commited the modifications.

**Comment by stefaweb** [comment]
Put VARIABLE in small cap. Easy to read.

$lang = array(
	// 01-login.php
	'username01' => 'Utilisateur',
	'password01' => 'Mot de passe',
	// 03-functions.php
	'jumpmessage03' => 'Aller au message :',
	'cuser03' => 'Utilisateur',
	'cst03' => 'Heure système',
	'colorcodes03' => 'Codes couleur',
	'ccodes01_03' => 'Contenu dangereux/infecté',
	'ccodes02_03' => 'Liste blanche',
	'ccodes03_03' => 'Liste noire',
	'ccodes04_03' => 'Non vérifié',
	'ccodes05_03' => 'Bon',

Than.

// 01-login.php
'USERNAME01' => 'Utilisateur',
'PASSWORD01' => 'Mot de passe',

// 03-functions.php
'JUMPMESSAGE03' => 'Aller au message :',
'CUSER03' => 'Utilisateur',
'CST03' => 'Heure système',
'COLORCODES03' => 'Codes couleur',
'CCODES01_03' => 'Contenu dangereux/Infecté',
'CCODES02_03' => 'Liste Blanche',
'CCODES03_03' => 'Liste Noire',
'CCODES04_03' => 'Non vérifié',
'CCODES05_03' => 'Bon',

**Comment by stefaweb** [comment]
But in fact my idea is wrong (small ca). Big Cap are a convention for variable.

Why the accents are html code? 

endelwar said ok for full UTF8 text.

**Comment by stefaweb** [comment]
Commited on my fork with mods.
https://github.com/stefaweb/1.2.0/blob/0db01d194e5939a8b2a7ee5767734c4e56c29459/mailscanner/languages/fr.php

Note this:
'i18_missing' => 'Non traduit en français'
'i18_missing' => 'Not translated.'

**Comment by endelwar** [comment]
Big letters are a standard for constants in php, small letter for variables

**Comment by stefaweb** [comment]
CONSTANTS as well. ;-)

**Comment by rafaelbsd** [comment]
Yeah, I kown... but I did the translation with constants the first time....
So it was easy to make de diference... now I´ll convert all to lower cases....

About the accents, old habits die hard.... I always used like that, so I never have to worry about encodings...

**Comment by rafaelbsd** [comment]
OK... files commited... everything is in lower case now..... ;-)

File: mailscanner/detail.php [hunk 1]
```diff
@@ -63,40 +63,40 @@
 // The sql command to pull the data
 $sql = "
  SELECT
-  DATE_FORMAT(timestamp, '" . DATE_FORMAT . " " . TIME_FORMAT . "') AS 'Received on:',
-  hostname AS 'Received by:',
-  clientip AS 'Received from:',
-  headers 'Received Via:',
+  date_format(timestamp, '" . date_format . " " . time_format . "') as '" . $lang['receivedon04'] . ":',
```
Comment (endelwar) [inline mailscanner/detail.php]: date_format and time_format must be uppercase

File: mailscanner/detail.php [hunk 2]
```diff
@@ -63,40 +63,40 @@
...
+  hostname as '" . $lang['receivedby04'] . ":',
+  clientip as '" . $lang['receivedfrom04'] . ":',
+  headers '" . $lang['receivedvia04'] . ":',
   id AS 'ID:',
-  headers AS 'Message Headers:',
-  from_address AS 'From:',
-  to_address AS 'To:',
-  subject AS 'Subject:',
-  size AS 'Size:',
+  headers as '" . $lang['msgheaders04'] . ":',
```
Comment (endelwar) [inline mailscanner/detail.php]: all MySQL reserved words should be in uppercase (AS, CASE, WHEN, THEN, ELSE....)

File: mailscanner/detail.php [hunk 3]
```diff
@@ -360,7 +361,7 @@
 if ((is_array($quarantined)) && (count($quarantined) > 0)) {
     echo "<br>\n";
 
-    if (isset($_GET['submit']) && ($_GET['submit'] == "Submit")) {
+    if (isset($_get['submit']) && ($_get['submit'] == $lang['submit04'])) {
```
Comment (endelwar) [inline mailscanner/detail.php]: $_GET must be uppercase

**Comment by rafaelbsd** [comment]
OH my... my script has converted more than it needs.... I´ll ajust that....

**Comment by rafaelbsd** [comment]
OK...

File: mailscanner/functions.php [hunk 4]
```diff
@@ -96,7 +100,7 @@
             define('VIRUS_REGEX', '/^Dummy$/');
             break;
         case 'sophos':
-            define('VIRUS_REGEX', '/(>>>) Virus \'(\S+)\' found/');
+            define('VIRUS_REGEX', '/(>>>) Virus \'(\S+)\' '.$ENCONTRADO03.'/');
```
Comment (endelwar) [inline mailscanner/functions.php]: this should be  $lang

File: mailscanner/login.php [hunk 5]
```diff
@@ -168,11 +170,11 @@
                     </p>
                 <?php } ?>
 
-                <p><label for="myusername">Username</label></p>
+                <p><label for="myusername"><?php echo $lang['username'] ?></label></p>
```
Comment (endelwar) [inline mailscanner/login.php]: please put a `;` at the end

File: mailscanner/login.php [hunk 6]
```diff
@@ -168,11 +170,11 @@
...
 
                 <p><input name="myusername" type="text" id="myusername" autofocus></p>
 
-                <p><label for="mypassword">Password</label></p>
+                <p><label for="mypassword"><?php echo $lang['password'] ?></label></p>
```
Comment (endelwar) [inline mailscanner/login.php]: please put a `;` at the end

File: mailscanner/status.php [hunk 7]
```diff
@@ -38,7 +38,7 @@
 session_start();
 require('login.function.php');
 
-$refresh = html_start("Recent Messages", STATUS_REFRESH, false, false);
+$refresh = html_start($lang['RECENTMESSAGES01'], STATUS_REFRESH, false, false);
```
Comment (endelwar) [inline mailscanner/status.php]: lowercase

File: mailscanner/functions.php [hunk 8]
```diff
@@ -96,7 +100,7 @@
             define('VIRUS_REGEX', '/^Dummy$/');
             break;
         case 'sophos':
-            define('VIRUS_REGEX', '/(>>>) Virus \'(\S+)\' found/');
+            define('VIRUS_REGEX', '/(>>>) Virus \'(\S+)\' '.$lang['found03'].'/');
```
Comment (endelwar) [inline mailscanner/functions.php]: I don't know if Sophos has localized messages, better keep "found" here

File: mailscanner/languages/en.php [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    'receivedvia04' => 'Received Via',
+    'msgheaders04' => 'Message Headers',
+    'from04' => 'From',
+    'to04' => 'To',
+    'size04' => 'Size',
+    'subject04' => 'Subject',
+    'hdrantivirus04' => 'Anti-Virus/Dangerous Content Protection',
+    'blkfile04' => 'Blocked File',
+    'otherinfec04' => 'Other Infection',
+    'otherinfec04' => 'Other Infection',
```
Comment (endelwar) [inline mailscanner/languages/en.php]: this line is duplicated

**Comment by stefaweb** [comment]
Good evening.

I added fr.php 0.0.8.
Also committed many changes in functions.php, sf_version.php, other.php, quarantine.php and others files. Check my last commits.

**Comment by stefaweb** [comment]
Hello!

Hello!

Concerning the translation.

Could you place the comma " : " in the loc file?

Sample:

In English, this is   "From:"
In French, this is    "De :"   (with a space before comma).

For this reason, we need to have the comma in the loc file.

**Comment by stefaweb** [comment]
Sorry, I was speaking of "colon" not "comma".

I also have a trouble with fr.php. Scrutinizer cry for a bug. Don't find the problem.

https://scrutinizer-ci.com/g/mailwatch/1.2.0/inspections/fd91492f-e40d-4aa6-82ce-ce9d71bc461c/issues/files/mailscanner/languages/fr.php?status=new&selectedLabels%5B0%5D=9&orderField=path&order=asc

**Comment by stefaweb** [comment]
In quarantine.php:

            db_colorised_table($sql, 'Folder: ' . translateQuarantineDate($dir), true, true);
        } else {
            echo "No quarantined messages found\n";

---

            db_colorised_table($sql, $lang['folder09'] . translateQuarantineDate($dir), true, true);
        } else {
            echo "No quarantined messages found\n";

---
'folder08' => 'Quarantine Folder',
'folder08' => 'Dossier de quarantaine',
'folder09' => 'Folder:',
'folder09' => 'Dossier :',

And "No quarantined messages found" to add.

**Comment by endelwar** [comment]
@stefaweb a comma is missing on row 129

**Comment by stefaweb** [comment]
Thanks.
Yes found. Correct in 0.1.1.

**Comment by rafaelbsd** [comment]
Hi... Just commited a few more translations. @stefaweb , I changed the variables to include ":"...

**Comment by stefaweb** [comment]
Hello @rafaelbsd!

I've updated here the last fr.php file (#191). Thanks for the ":".

I've some comments:

-------
In quarantine.php:

We need a second "Folder" translation here as in French the words are not in the same direction that English.

        db_colorised_table($sql, 'Folder: ' . translateQuarantineDate($dir), true, true);
    } else {
        echo "No quarantined messages found\n";
        db_colorised_table($sql, $lang['folder09'] . translateQuarantineDate($dir), true, true);
    } else {
        echo "No quarantined messages found\n";
'folder08' => 'Quarantine Folder',
'folder08' => 'Dossier de quarantaine',
'folder09' => 'Folder:',
'folder09' => 'Dossier :',

------

in en.php (and pt).

'to0203' => 'to�',

What is the special character found in the to0203 variable?

-----------

It will be nice to have a system to manage missing translation.
'i18_missing' => 'Non traduit en français',

Cheers,

File: mailscanner/functions.php [hunk 10]
```diff
@@ -244,29 +249,29 @@ function html_start($title, $refresh = 0, $cacheable = true, $report = false)
...
     echo '<tr><td>' . $_SESSION['fullname'] . '</td><td><span id="clock">&nbsp;</span></td></tr>' . "\n";
     echo '</table>' . "\n";
     echo '</td>' . "\n";
 
     echo '<td align="left" valign="top">' . "\n";
     echo '   <table border="0" cellpadding="1" cellspacing="1" class="mail">' . "\n";
-    echo '    <tr> <th colspan="2">Color Codes</th> </tr>' . "\n";
-    echo '    <tr> <td>Bad Content/Infected</TD> <td class="infected"></TD> </TR>' . "\n";
+    echo '    <tr> <th colspan="2">' . $lang['colorcodes03'] . '</th> </tr>' . "\n";
+    echo '    <tr> <td>' . $lang['ccodes01_03'] . '</TD> <td class="infected"></TD> </TR>' . "\n";
```
Comment (endelwar) [inline mailscanner/functions.php]: This variable name is not self explanatory, should be `$lang['badcontentinfected_03']`

File: mailscanner/functions.php [hunk 11]
```diff
@@ -244,29 +249,29 @@ function html_start($title, $refresh = 0, $cacheable = true, $report = false)
...
     echo '    <tr> <td>High Spam</td> <td class="highspam"></td> </tr>' . "\n";
     if (get_conf_truefalse('mcpchecks')) {
         echo '    <tr> <td>MCP</td> <td class="mcp"></td> </tr>' . "\n";
         echo '    <tr> <td>High MCP</td><td class="highmcp"></td></tr>' . "\n";
     }
-    echo '    <tr> <td>Whitelisted</td> <td class="whitelisted"></td> </tr>' . "\n";
-    echo '    <tr> <td>Blacklisted</td> <td class="blacklisted"></td> </tr>' . "\n";
-    echo '	  <tr> <td>Not Scanned</td> <td class="notscanned"></td> </tr>' . "\n";
-    echo '    <tr> <td>Clean</td> <td></td> </tr>' . "\n";
+    echo '    <tr> <td>' . $lang['ccodes02_03'] . '</td> <td class="whitelisted"></td> </tr>' . "\n";
```
Comment (endelwar) [inline mailscanner/functions.php]: This variable name is not self explanatory, should be `$lang['whitelisted_03']`

File: mailscanner/functions.php [hunk 12]
```diff
@@ -244,29 +249,29 @@ function html_start($title, $refresh = 0, $cacheable = true, $report = false)
...
+    echo '    <tr> <td>' . $lang['ccodes03_03'] . '</td> <td class="blacklisted"></td> </tr>' . "\n";
```
Comment (endelwar) [inline mailscanner/functions.php]: This variable name is not self explanatory, should be `$lang['blacklisted_03']`

File: mailscanner/functions.php [hunk 13]
```diff
@@ -244,29 +249,29 @@ function html_start($title, $refresh = 0, $cacheable = true, $report = false)
...
+    echo '        <tr> <td>' . $lang['ccodes04_03'] . '</td> <td class="notscanned"></td> </tr>' . "\n";
```
Comment (endelwar) [inline mailscanner/functions.php]: This variable name is not self explanatory, should be `$lang['notscanned_03']`

File: mailscanner/functions.php [hunk 14]
```diff
@@ -244,29 +249,29 @@ function html_start($title, $refresh = 0, $cacheable = true, $report = false)
...
+    echo '    <tr> <td>' . $lang['ccodes05_03'] . '</td> <td></td> </tr>' . "\n";
```
Comment (endelwar) [inline mailscanner/functions.php]: This variable name is not self explanatory, should be `$lang['clean_03']`

**Comment by endelwar** [comment]
@rafaelbsd this is starting to get a good shape, well done!
If you fix some of the comments that I left in some source lines I'll merge this PR.

After the merge I'll implement a function as explained here https://github.com/mailwatch/1.2.0/pull/170#issuecomment-113507492 and will call for help in users mailing-list for translations

**Comment by rafaelbsd** [comment]
Hi @stefaweb , I created a new variable in quarantine.php:
'folder_0208' => 'Folder' (the last number is a reference to the file, to help other translators...), in the place that you´ve asked.

**Comment by rafaelbsd** [comment]
Hi @endelwar , I changed the variable names to the ones you´ve suggested. I´l now work on integrate your last changes into my code.

**Comment by stefaweb** [comment]
@rafaelbsd 
Thanks. I will change that in fr.php.

**Comment by stefaweb** [comment]
@rafaelbsd 

in en.php

----
'badcontentinfected03' => 'Bad Contect/Infected',
Content or Context?
Content for me.
----
'others03' => 'Otherd',
Otherd? May be "Other".
----
'seconds05' => 'segunds',
segunds?
May be "second(s)"
----
'to0203' => 'to©',
©?
----

**Comment by rafaelbsd** [comment]
Hi, @stefaweb . Sorry about that ;-)... Already fixed. Thanks!

**Comment by rafaelbsd** [comment]
Hi, @endelwar . I made a review using github comparison, and didn´t found any differences comparing masters, except, of course, what has changed for the translations. Am I missing something?

Thanks!

**Comment by stefaweb** [comment]
I'm not the only one to suffer with Git. ;-)

Try this:

http://blog.michelemattioni.me/2013/01/29/what-do-i-do-when-my-pull-request-does-not-merge-automatically-in-master/

**Comment by endelwar** [comment]
@rafaelbsd as noted by Stéphane the automatic merge cannot be done if you don't work on a new branch instead of master branch.
Don't worry, I can work this around by merging the .patch file generated by github automatically (https://patch-diff.githubusercontent.com/raw/mailwatch/1.2.0/pull/170.patch)

**Comment by stefaweb** [comment]
I did the same error, working directly on the forked master branch...

**Comment by rafaelbsd** [comment]
Yes, I just start to work with GitHub for this project. I´ll create a different branch....

**Pull request closed without merging**